MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs

```diff
@@ -32,323 +28,288 @@ namespace MediaBrowser.MediaEncoding.Encoder
+                FFprobePath = path;
+                return true;
             }
 
-            return new Tuple<string, string>(null, null);
+            return false;
         }
 
-        private Tuple<string, string> TestForInstalledVersions()
+        private string GetEncoderPath(string path)
         {
-            string encoderPath = null;
-            string probePath = null;
-
-            if (_hasExternalEncoder && ValidateVersion(_originalFFMpegPath, true))
+            if (Directory.Exists(path))
             {
-                encoderPath = _originalFFMpegPath;
-                probePath = _originalFFProbePath;
+                return GetEncoderPathFromDirectory(path);
             }
 
-            if (string.IsNullOrWhiteSpace(encoderPath))
+            if (File.Exists(path))
             {
-                if (ValidateVersion("ffmpeg", true) && ValidateVersion("ffprobe", false))
-                {
-                    encoderPath = "ffmpeg";
-                    probePath = "ffprobe";
-                }
+                return path;
             }
 
-            return new Tuple<string, string>(encoderPath, probePath);
+            return null;
         }
 
-        private Tuple<string, string> GetPathsFromDirectory(string path)
+        private string GetEncoderPathFromDirectory(string path)
         {
-            // Since we can't predict the file extension, first try directly within the folder
-            // If that doesn't pan out, then do a recursive search
-            var files = FileSystem.GetFilePaths(path);
-
-            var excludeExtensions = new[] { ".c" };
-
-            var ffmpegPath = files.FirstOrDefault(i => string.Equals(Path.GetFileNameWithoutExtension(i), "ffmpeg", StringComparison.OrdinalIgnoreCase) && !excludeExtensions.Contains(Path.GetExtension(i) ?? string.Empty));
-            var ffprobePath = files.FirstOrDefault(i => string.Equals(Path.GetFileNameWithoutExtension(i), "ffprobe", StringComparison.OrdinalIgnoreCase) && !excludeExtensions.Contains(Path.GetExtension(i) ?? string.Empty));
-
-            if (string.IsNullOrWhiteSpace(ffmpegPath) || !File.Exists(ffmpegPath))
+            try
             {
-                files = FileSystem.GetFilePaths(path, true);
+                var files = FileSystem.GetFilePaths(path);
 
-                ffmpegPath = files.FirstOrDefault(i => string.Equals(Path.GetFileNameWithoutExtension(i), "ffmpeg", StringComparison.OrdinalIgnoreCase) && !excludeExtensions.Contains(Path.GetExtension(i) ?? string.Empty));
+                var excludeExtensions = new[] { ".c" };
 
-                if (!string.IsNullOrWhiteSpace(ffmpegPath))
-                {
-                    ffprobePath = GetProbePathFromEncoderPath(ffmpegPath);
-                }
+                return files.FirstOrDefault(i => string.Equals(Path.GetFileNameWithoutExtension(i), "ffmpeg", StringComparison.OrdinalIgnoreCase) && !excludeExtensions.Contains(Path.GetExtension(i) ?? string.Empty));
```

Comment:
what is this?.. why not _at least_ check that found path is executable? otherwise having a `ffmpeg.txt` next to `ffmpeg` itself somewhere could break everything

Reply:
How do you check for executable on Windows?

Reply:
No idea how to do in C# :) Something checking for `FileSystemRights.ExecuteFile` presence somewhere in file security attributes probably?

Reply:
All, I'm looking for suggestions on how to improve this please.